### PR TITLE
[REVIEW] Change dataframe to call new join api

### PR DESCRIPTION
**Pull request by dantegd**
Changed join and merge in dataframe to use new libgdf join API: 

https://github.com/gpuopenanalytics/libgdf/pull/135

New join builds the full dataframe in libgdf/C++ instead of only calculating indices in libgdf, providing a big speedup. 

Note: tests need to be rerun after libgdf PR 135 merge.

### CODE REVIEW

File: pygdf/_gdf.py [hunk 1]
```diff
@@ -132,6 +132,22 @@ def np_to_gdf_dtype(dtype):
...
+    """
+    return np.dtype({
+         libgdf.GDF_FLOAT64: np.float64,
+         libgdf.GDF_FLOAT32: np.float32,
+         libgdf.GDF_INT64: np.int64,
+         libgdf.GDF_INT32: np.int32,
+         libgdf.GDF_INT16: np.int16,
+         libgdf.GDF_INT8: np.int8,
+         libgdf.GDF_INT8: np.bool_,
+         libgdf.GDF_DATE64: np.datetime64,
```
Comment (kkraus14) [inline pygdf/_gdf.py]: Will this not have the proper units of "ms"?

File: pygdf/_gdf.py [hunk 2]
```diff
@@ -239,6 +255,96 @@ def apply_join(col_lhs, col_rhs, how, method='hash'):
...
+        libgdf.gdf_context_view(gdf_context, 0, method_api, 0)
+    elif method == 'sort':
+        libgdf.gdf_context_view(gdf_context, 1, method_api, 0)
+    else:
+        msg = "method not supported"
+        raise ValueError(msg)
+
+    if(how not in ['left', 'inner']):
+        msg = "new join api must be left or inner"
+        raise ValueError(msg)
```
Comment (kkraus14) [inline pygdf/_gdf.py]: Not sure if this check is redundant, but leave it in for now to be safe.

File: pygdf/dataframe.py [hunk 3]
```diff
@@ -712,40 +712,39 @@ def merge(self, other, on=None, how='left', lsuffix='_x', rsuffix='_y',
...
             if name in same_names:
                 return "{}{}".format(name, suffix)
             return name
 
-        def gather_cols(outdf, indf, on, idx, joinidx, suffix):
-            mask = (Series(idx) != -1).as_mask()
-            for k in on:
-                newcol = indf[k].take(idx).set_mask(mask).set_index(joinidx)
-                outdf[fix_name(k, suffix)] = newcol
+        # Columns are returned in order left - on - right fromm libgdf
```
Comment (kkraus14) [inline pygdf/dataframe.py]: Typo, extra m.

File: pygdf/dataframe.py [hunk 4]
```diff
@@ -712,40 +712,39 @@ def merge(self, other, on=None, how='left', lsuffix='_x', rsuffix='_y',
...
+        # Creating dataframe with ordering as pandas:
 
-        def gather_empty(outdf, indf, idx, joinidx, suffix):
-            for k in indf.columns:
-                outdf[fix_name(k, suffix)] = indf[k][:0]
+        gap = len(self.columns) - len(on)
+        for idx in range(len(on)):
+            df[on[idx]] = cols[idx + gap]
+            print(idx)
```
Comment (kkraus14) [inline pygdf/dataframe.py]: Will need to remove this print statement 😄

File: pygdf/dataframe.py [hunk 5]
```diff
@@ -712,40 +712,39 @@ def merge(self, other, on=None, how='left', lsuffix='_x', rsuffix='_y',
...
-                    left_indices, df.index, lsuffix)
-        gather_cols(df, rhs, [x for x in rhs.columns if x not in on],
-                    right_indices, df.index, rsuffix)
+        idx = 0
+
+        for name in self.columns:
+            if name not in on:
+                df[fix_name(name, lsuffix)] = cols[idx]
+                idx = idx + 1
+                print(idx)
```
Comment (kkraus14) [inline pygdf/dataframe.py]: This one too 😄

File: pygdf/dataframe.py [hunk 6]
```diff
@@ -712,40 +712,39 @@ def merge(self, other, on=None, how='left', lsuffix='_x', rsuffix='_y',
...
+
+        idx = len(self.columns)
+
+        for name in other.columns:
+            if name not in on:
+                df[fix_name(name, rsuffix)] = cols[idx]
+                idx = idx + 1
+                print(idx)
```
Comment (kkraus14) [inline pygdf/dataframe.py]: And one more 😄

File: pygdf/_gdf.py [hunk 7]
```diff
@@ -132,6 +134,22 @@ def np_to_gdf_dtype(dtype):
...
+    """Util to convert gdf dtype to numpy dtype.
+    """
+    return np.dtype({
+         libgdf.GDF_FLOAT64: np.float64,
+         libgdf.GDF_FLOAT32: np.float32,
+         libgdf.GDF_INT64: np.int64,
+         libgdf.GDF_INT32: np.int32,
+         libgdf.GDF_INT16: np.int16,
+         libgdf.GDF_INT8: np.int8,
+         libgdf.GDF_INT8: np.bool_,
```
Comment (kkraus14) [inline pygdf/_gdf.py]: This is going to overwrite the libgdf.GDF_INT8 key.

File: pygdf/_gdf.py [hunk 8]
```diff
@@ -132,6 +134,21 @@ def np_to_gdf_dtype(dtype):
...
+    return np.dtype({
+         libgdf.GDF_FLOAT64: np.float64,
+         libgdf.GDF_FLOAT32: np.float32,
+         libgdf.GDF_INT64: np.int64,
+         libgdf.GDF_INT32: np.int32,
+         libgdf.GDF_INT16: np.int16,
+         libgdf.GDF_INT8: np.int8,
+         libgdf.GDF_DATE64: np.dtype('datetime64[ms]'),
+         libgdf.N_GDF_TYPES: np.int32,
+     }[dtype])
```
Comment (scopatz) [inline pygdf/_gdf.py]: This dict should be defined at module-level scope, rather than being reinstantiated each call.

File: pygdf/_gdf.py [hunk 9]
```diff
@@ -239,6 +256,92 @@ def apply_join(col_lhs, col_rhs, how, method='hash'):
...
+
+    if method == 'hash':
+        libgdf.gdf_context_view(gdf_context, 0, method_api, 0)
+    elif method == 'sort':
+        libgdf.gdf_context_view(gdf_context, 1, method_api, 0)
+    else:
+        msg = "method not supported"
+        raise ValueError(msg)
+
+    if(how not in ['left', 'inner']):
```
Comment (scopatz) [inline pygdf/_gdf.py]: PEP8, there should be a space after the `if`

File: pygdf/_gdf.py [hunk 10]
```diff
@@ -239,6 +256,92 @@ def apply_join(col_lhs, col_rhs, how, method='hash'):
...
+        else:
+            left_idx.append(idx)
+        idx = idx + 1
+
+    for name in on:
+        result_cols.append(columnview(0, None,
+                                      dtype=col_lhs[name]._column.dtype))
+        result_col_names.append(name)
+
+    # right_idx = ffi.new('int[' + str(len(on)) + ']')
```
Comment (scopatz) [inline pygdf/_gdf.py]: Please remove dead code

File: pygdf/_gdf.py [hunk 11]
```diff
@@ -15,6 +15,8 @@
 from . import cudautils
 from .utils import mask_bitsize
 
+mask_dtype = np.dtype(np.uint8)
```
Comment (scopatz) [inline pygdf/_gdf.py]: PEP8, module-level variables should be in ALL_UPPERCASE

**Comment by scopatz** [comment]
This seems like it could use some tests

File: pygdf/_gdf.py [hunk 12]
```diff
@@ -239,6 +256,92 @@ def apply_join(col_lhs, col_rhs, how, method='hash'):
...
+    right_idx = []
+    idx = 0
+    for name, col in col_rhs.items():
+        list_rhs.append(col._column.cffi_view)
+
+        if name not in on:
+            result_cols.append(columnview(0, None, dtype=col._column.dtype))
+            result_col_names.append(name)
+        else:
+            right_idx.append(idx)
```
Comment (kkraus14) [inline pygdf/_gdf.py]: Does the libgdf call expect the indices to be the same for the matching columns? I.E. left_idx[0] and right_idx[0] correspond to the matching column.

If so, if the columns aren't in the same order this won't match up properly.

File: pygdf/dataframe.py [hunk 13]
```diff
@@ -665,7 +668,7 @@ def _n_largest_or_smallest(self, method, n, columns, keep):
         return df.set_index(self.index.take(new_positions))
 
     def merge(self, other, on=None, how='left', lsuffix='_x', rsuffix='_y',
-              type='sort'):
+              type='hash', method='hash'):
```
Comment (kkraus14) [inline pygdf/dataframe.py]: Don't need the type parameter anymore if this is being fixed.

**Comment by kkraus14** [comment]
rerun tests

**Comment by kkraus14** [comment]
LGTM pending CI passing

**Pull request merged**